### PR TITLE
refactor: hide polynomial math behind SecretPolynomial helper (#11)

### LINKED ISSUE #11: Refactor: hide polynomial math behind a SecretPolynomial helper
## Background

`ShamirSecretSharing/ShamirSecretSharingService.cs` smears polynomial-shaped concerns across two methods without ever naming a `Polynomial` concept:

- `SplitSecret` (lines 60-94) allocates a length-`t` `int[] coefficients` per secret byte, runs a private `EvaluatePolynomial(int[], int)` helper, and evaluates at `x = 1..n` inline.
- `ReconstructSecret` (lines 134-167) open-codes Lagrange interpolation at `x = 0` with ~7 primitive `_field.Multiply`/`Subtract`/`Divide` calls per inner-loop iteration. The polynomial is implicit.
- The private `EvaluatePolynomial` (lines 176-187) is pass-through-shaped — `(int[], int) -> int` — and only used from one site.

Reading either method requires reverse-engineering "this is `L_i(0)`" from raw modular arithmetic. The concept never gets a name, and the call sites stay coupled to `FiniteField`'s primitive ops.

## Proposed interface (minimal design)

Hide `Polynomial` entirely. Two static methods on an internal helper:

```csharp
internal static class SecretPolynomial
{
    /// Samples a random degree-(threshold-1) polynomial with constant term `secretByte`
    /// over `field`, then evaluates it at each x in `xs`. Returns ys[i] = P(xs[i]).
    public static int[] SampleAndEvaluate(
        int secretByte,
        int threshold,
        ReadOnlySpan<int> xs,
        FiniteField field,
        RandomNumberGenerator rng);

    /// Lagrange-interpolates P(0) over `field` from the points (xs[i], ys[i]).
    public static int InterpolateAtZero(
        ReadOnlySpan<int> xs,
        ReadOnlySpan<int> ys,
        FiniteField field);
}
```

No `Polynomial` type leaks. Pure / stateless / thread-safe.

## Call-site impact

`SplitSecret` inner loop:

```csharp
int[] xs = Enumerable.Range(1, n).ToArray();
for (int b = 0; b < secret.Length; b++)
{
    int[] ys = SecretPolynomial.SampleAndEvaluate(secret[b], t, xs, _field, _rng);
    for (int i = 0; i < n; i++) yValuesPerShare[i][b] = ys[i];
}
```

`ReconstructSecret` inner loop:

```csharp
for (int b = 0; b < secretLength; b++)
{
    for (int i = 0; i < shares.Count; i++) ys[i] = distinctShares[i].YValues[b];
    secret[b] = (byte)SecretPolynomial.InterpolateAtZero(xs, ys, _field);
}
```

Zero `_field.*` calls at the SSS service level. Zero coefficient arrays in scope. The two ~30-line inner loops collapse to ~5 lines each.

## Dependency strategy

- `FiniteField` and `RandomNumberGenerator` passed as parameters per call.
- `SecretPolynomial` is `internal` plus `InternalsVisibleTo` for the test project so tests can target the helper directly.

## Test impact

New tests (replace implicit round-trip coverage):

- `SecretPolynomial.SampleAndEvaluate` produces a degree-`t-1` polynomial whose `P(0) == secretByte` (interpolate back and assert).
- Determinism with a scripted `RandomNumberGenerator` (mock).
- `InterpolateAtZero` against published SSS vectors.
- Boundary cases: `t == 2`, `t == n`, all-zero coefficients, secret byte 0 / 255.

Existing `FiniteField_*` primitive tests stay — still pull their weight.

## Risks / trade-offs

- Forbids polynomial reuse, interpolation at `x != 0`, and coefficient inspection. Acceptable for SSS — every byte uses a fresh polynomial and reconstruction only ever interpolates at 0.
- No incremental share addition. Acceptable: callers know `n` upfront.
- Per-byte allocation of `ys` array; pool if profiling shows hot.

## Open questions

- Should `SampleAndEvaluate` accept the random source as `IFieldRandomSource` (the seam from #issue-C2) instead of `RandomNumberGenerator`? If C2 lands first, yes. If this lands first, parameterize as `RandomNumberGenerator` and migrate when C2 ships.
- Naming: `SecretPolynomial` vs `ShamirPolynomial` vs just `Polynomial`. Leaning `SecretPolynomial` because it is SSS-specific.

## Related

This is the first of four module-deepening RFCs identified in a `/improve-codebase-architecture` pass. See companion issues for the IRandomSource seam, Splitter/Combiner separation, and FiniteField primality factory.

### PR DESCRIPTION
**Pull request by stevehansen**
## Summary
- Extracts polynomial sampling/evaluation and Lagrange interpolation into a new `internal static class SecretPolynomial` with two methods: `SampleAndEvaluate(secretByte, threshold, xs, field, rng) -> int[]` and `InterpolateAtZero(xs, ys, field) -> int`.
- `ShamirSecretSharingService` is fully decoupled from `FiniteField` primitives — zero `_field.{Add,Subtract,Multiply,Divide}` calls remain in the service; the private `EvaluatePolynomial` helper is gone; the two ~30-line inner loops collapsed to ~5 lines each.
- Service is now `IDisposable` since it owns a `RandomNumberGenerator` instance (previously it relied on the static `RandomNumberGenerator.Fill`).
- `InterpolateAtZero` validates `xs.Length == ys.Length`.
- `InternalsVisibleTo("ShamirSecretSharingTests")` added so tests can target the helper directly.

Closes #11.

## Test plan
- [x] `dotnet build` — clean (0 warnings, 0 errors; `TreatWarningsAsErrors=true` honored across net8.0 and net9.0).
- [x] `dotnet test` — 39 passed, 0 failed (28 pre-existing + 11 new in `SecretPolynomialTests.cs`).
- [x] New tests cover: round-trip across (t,n) combos, scripted-RNG determinism, hand-computed known vectors (linear y=3x+2; quadratic 7+2x+x²; P(x)=10+3x+3x² → ys=[16,28,46]), any-threshold-subset recovery, boundary secrets (0, 255), minimum threshold = 2.
- [x] Verified zero `_field.*` arithmetic calls in service (only `_field.Prime` for validation).
- [x] Public API unchanged — signatures, exception types, and validation messages match pre-refactor exactly.

## Notes for reviewer
The diff base (`development`) is currently 3 commits behind on the remote, so this PR's view will include those prior docs/Share fix commits in addition to the refactor. The issue #11 work itself is in commit `1190701` only.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: ShamirSecretSharing/SecretPolynomial.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            var powerOfX = 1;
+            foreach (var coeff in coefficients)
+            {
+                result = field.Add(result, field.Multiply(coeff, powerOfX));
+                powerOfX = field.Multiply(powerOfX, x);
+            }
+            ys[i] = result;
+        }
+        return ys;
+    }
```
[the remark this answers is not in the document]
Comment (stevehansen) [reply ShamirSecretSharing/SecretPolynomial.cs]: Applied in 7f2753e. Switched to RandomNumberGenerator.GetInt32(Prime) for unbiased coefficients, stackalloc'd the coefficient buffer, and changed the API to write into a caller-supplied destination span — no per-byte byte[] or int[] allocations now.

File: ShamirSecretSharing/ShamirSecretSharingService.cs [hunk 2]
```diff
@@ -128,69 +132,28 @@ public byte[] ReconstructSecret(IReadOnlyList<Share> shares, int t)
...
-    {
-        var result = 0;
-        var powerOfX = 1; // x^0
-        foreach (var coeff in coefficients)
-        {
-            var term = _field.Multiply(coeff, powerOfX);
-            result = _field.Add(result, term);
-            powerOfX = _field.Multiply(powerOfX, x);
+            reconstructedSecret[byteIndex] = (byte)SecretPolynomial.InterpolateAtZero(xs, ys, _field);
         }
```
[the remark this answers is not in the document]
Comment (stevehansen) [reply ShamirSecretSharing/ShamirSecretSharingService.cs]: Applied in 7f2753e. ReconstructSecret precomputes the basis once via ComputeLagrangeBasisAtZero outside the byte loop; per-byte work is now O(t).

File: ShamirSecretSharing/SecretPolynomial.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                if (i == j) continue;
+                var xj = xs[j];
+                numerator = field.Multiply(numerator, xj);
+                denominator = field.Multiply(denominator, field.Subtract(xj, xi));
+            }
+            var basis = field.Divide(numerator, denominator);
+            result = field.Add(result, field.Multiply(ys[i], basis));
+        }
+        return result;
+    }
```
[the remark this answers is not in the document]
Comment (stevehansen) [reply ShamirSecretSharing/SecretPolynomial.cs]: Applied in 7f2753e. Split into ComputeLagrangeBasisAtZero + InterpolateWithBasis; InterpolateAtZero remains as a composition convenience for single-shot interpolation.

File: ShamirSecretSharing/ShamirSecretSharingService.cs [hunk 4]
```diff
@@ -9,34 +9,55 @@ namespace ShamirSecretSharing;
...
+
+    /// <summary>
+    /// Releases the unmanaged resources used by this instance and optionally the managed resources.
+    /// </summary>
+    /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
+    protected virtual void Dispose(bool disposing)
+    {
+        if (disposing)
+            _rng.Dispose();
     }
```
[the remark this answers is not in the document]
Comment (stevehansen) [reply ShamirSecretSharing/ShamirSecretSharingService.cs]: Applied in 7f2753e. Dropped IDisposable and the _rng field. SecretPolynomial now uses the static RandomNumberGenerator.GetInt32 — no instance to manage, and it fixes the bias issue in the same move.

File: ShamirSecretSharing/ShamirSecretSharingService.cs [hunk 5]
```diff
@@ -52,49 +73,32 @@ public Share[] SplitSecret(byte[] secret, int n, int t)
...
-            // Generate n points (shares) on this polynomial
-            for (var shareIndex = 0; shareIndex < n; shareIndex++)
-            {
-                var x = shareIndex + 1; // x-coordinates are 1, 2, ..., n (must be non-zero)
-                var y = EvaluatePolynomial(coefficients, x);
-                yValuesPerShare[shareIndex][byteIndex] = y;
-            }
+            var ys = SecretPolynomial.SampleAndEvaluate(secretByte, t, xs, _field, _rng);
+            for (var i = 0; i < n; i++)
+                yValuesPerShare[i][byteIndex] = ys[i];
```
[the remark this answers is not in the document]
Comment (stevehansen) [reply ShamirSecretSharing/ShamirSecretSharingService.cs]: Applied in 7f2753e. ys is now a stackalloc'd destination span (with a heap fallback when n exceeds the stackalloc threshold).

File: CLAUDE.md [hunk 6]
```diff
@@ -4,65 +4,76 @@ This file provides guidance to Claude Code (claude.ai/code) when working with co
...
-- Uses `System.Security.Cryptography.RandomNumberGenerator` for cryptographic randomness
-- `Directory.Build.props` enforces `TreatWarningsAsErrors`, `EnforceCodeStyleInBuild`, nullable reference types, and latest C# language version across all projects
-- `GenerateDocumentationFile` is enabled on the library project, so all public members require XML doc comments
-- Renovate is configured for automated dependency updates
+- Default prime is **GF(257)** (smallest prime > 255), so each secret byte fits in one field element. With this prime, `n` (total shares) is capped at 256.
+- `(t, n)` threshold scheme — `t` shares are required and sufficient to reconstruct. `t` must be ≥ 2.
+- Coefficients are sampled with `System.Security.Cryptography.RandomNumberGenerator`. **Per-coefficient bias:** the implementation does `randomBytes[i] % Prime` on a single byte; for `Prime = 257` this is effectively uniform on 0–255 (value 256 is unreachable but the constant term is the secret byte, so the polynomial's other coefficients have ~1/256 bias — acceptable for the default prime but worth keeping in mind if you change the prime).
```
[the remark this answers is not in the document]
Comment (stevehansen) [reply CLAUDE.md]: Applied in 93ddf32. The bias paragraph in CLAUDE.md is gone — sampling is now via the unbiased RandomNumberGenerator.GetInt32(Prime).

File: ShamirSecretSharing/SecretPolynomial.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        for (var i = 0; i < xs.Length; i++)
+        {
+            var x = xs[i];
+            var result = 0;
+            var powerOfX = 1;
+            foreach (var coeff in coefficients)
+            {
+                result = field.Add(result, field.Multiply(coeff, powerOfX));
+                powerOfX = field.Multiply(powerOfX, x);
+            }
```
[the remark this answers is not in the document]
Comment (stevehansen) [reply ShamirSecretSharing/SecretPolynomial.cs]: Applied in 93ddf32. SampleAndEvaluate now uses Horner's method — t field multiplications per point instead of 2t.

File: ShamirSecretSharing/ShamirSecretSharingService.cs [hunk 8]
```diff
@@ -52,49 +52,34 @@ public Share[] SplitSecret(byte[] secret, int n, int t)
...
+            throw new ArgumentOutOfRangeException(nameof(n), $"Total shares n must be less than the prime modulus ({_field.Prime}).");
 
-        var shares = new Share[n];
         var yValuesPerShare = new int[n][];
         for (var i = 0; i < n; i++)
             yValuesPerShare[i] = new int[secret.Length];
 
-        // For each byte of the secret
-        var randomBytes = new byte[t - 1];
+        var xs = new int[n];
```
[the remark this answers is not in the document]
Comment (stevehansen) [reply ShamirSecretSharing/ShamirSecretSharingService.cs]: Applied in f2092fd. xs in SplitSecret is now a stackalloc'd Span<int> with a heap fallback above the threshold.

File: ShamirSecretSharing/ShamirSecretSharingService.cs [hunk 9]
```diff
@@ -128,69 +113,29 @@ public byte[] ReconstructSecret(IReadOnlyList<Share> shares, int t)
         if (distinctShares.Count < t)
             throw new ArgumentException($"Not enough distinct shares provided. Need {t} distinct X values, got {distinctShares.Count}.", nameof(shares));
 
+        var xs = new int[t];
```
[the remark this answers is not in the document]
Comment (stevehansen) [reply ShamirSecretSharing/ShamirSecretSharingService.cs]: Applied in f2092fd. xs in ReconstructSecret is now stackalloc'd too.

File: ShamirSecretSharing/SecretPolynomial.cs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        for (var i = 0; i < xs.Length; i++)
+        {
+            var x = xs[i];
+            // Horner's method: c0 + x*(c1 + x*(c2 + ... + x*c_{t-1})).
+            var result = 0;
+            for (var j = coefficients.Length - 1; j >= 0; j--)
+                result = field.Add(field.Multiply(result, x), coefficients[j]);
+            ys[i] = result;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (stevehansen) [reply ShamirSecretSharing/SecretPolynomial.cs:46]: Declining for now. The coefficients buffer is already stackalloc'd whenever threshold ≤ 64; the heap fallback only fires above that. Practical SSS thresholds are in the low double digits, so the fallback branch is effectively dead code today. Adding a `Span<int> coefficients` parameter to the helper's signature to optimize a path no realistic configuration hits is the kind of speculative flexibility this codebase avoids. Happy to revisit if a real use case ever exceeds the 64-coefficient threshold.

File: ShamirSecretSharing/ShamirSecretSharingService.cs [hunk 11]
```diff
@@ -52,49 +52,34 @@ public Share[] SplitSecret(byte[] secret, int n, int t)
...
-            for (var shareIndex = 0; shareIndex < n; shareIndex++)
-            {
-                var x = shareIndex + 1; // x-coordinates are 1, 2, ..., n (must be non-zero)
-                var y = EvaluatePolynomial(coefficients, x);
-                yValuesPerShare[shareIndex][byteIndex] = y;
-            }
+            SecretPolynomial.SampleAndEvaluate(secretByte, t, xs, ys, _field);
+            for (var i = 0; i < n; i++)
+                yValuesPerShare[i][byteIndex] = ys[i];
         }
```
[the remark this answers is not in the document]
Comment (stevehansen) [reply ShamirSecretSharing/ShamirSecretSharingService.cs:78]: Declining — pair of the SecretPolynomial.cs comment above. Same rationale: the coefficients stackalloc inside the helper already covers every realistic threshold; threading a pre-allocated buffer through the public-via-internal API costs more in clutter than it saves in (theoretical) allocations.
